lib/utils/fakeconn.go

```diff
@@ -201,6 +198,64 @@ func (c *ChConn) SetWriteDeadline(t time.Time) error {
 	return nil
 }
 
+// CancelableChConn is a wrapped SSH channel connection that supports deadlines.
+type CancelableChConn struct {
+	// ChConn is the base wrapped SSH channel connection.
+	*ChConn
+	// reader is the part of the pipe that clients read from.
+	reader net.Conn
+	// writer is the part of the pipe that receives data from SSH channel.
+	writer net.Conn
+}
+
+// NewCancelableChConn returns a new instance of wrapped SSH channel connection
+// that supports deadlines.
+func NewCancelableChConn(conn ssh.Conn, ch ssh.Channel) *CancelableChConn {
+	reader, writer := net.Pipe()
+	c := &CancelableChConn{
+		ChConn: NewChConn(conn, ch),
+		reader: reader,
+		writer: writer,
+	}
+	// Start copying from the SSH channel to the writer part of the pipe. The
+	// clients are reading from the reader part of the pipe (see Read below).
+	//
+	// This goroutine stops when either the SSH channel closes or this
+	// connection is closed e.g. by a http.Server (see Close below).
+	go io.Copy(writer, ch)
+	return c
+}
+
+// Read reads from the channel.
+func (c *CancelableChConn) Read(data []byte) (int, error) {
+	return c.reader.Read(data)
+}
+
+// SetDeadline sets the channel connection read/write deadlines.
+func (c *CancelableChConn) SetDeadline(t time.Time) error {
+	return c.reader.SetDeadline(t)
+}
+
+// SetReadDeadline sets the channel connection read deadline.
+func (c *CancelableChConn) SetReadDeadline(t time.Time) error {
+	return c.reader.SetReadDeadline(t)
+}
+
+// Closes closes all parts of the connection.
+func (c *CancelableChConn) Close() error {
+	var errors []error
+	if err := c.ChConn.Close(); err != nil {
+		errors = append(errors, err)
+	}
+	if err := c.reader.Close(); err != nil {
+		errors = append(errors, err)
+	}
+	if err := c.writer.Close(); err != nil {
+		errors = append(errors, err)
+	}
+	return trace.NewAggregate(errors...)
```

Comment:
```go
return trace.NewAggregate(c.ChConn.Close(), c.reader.Close(), c.writer.Close())
```

Reply:
I prefer the more verbose version in this case TBH.